src/zip.c

```diff
@@ -1038,3 +1074,395 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
   CLEANUP(zip);
   return NULL;
 }
+
+static mz_bool file_name_matches(const char *file_name,
+                                 const char *delete_name) {
+  int delete_name_length = strlen(delete_name);
+  char *delete_entry_name = strrpl(delete_name, delete_name_length, '\\', '/');
+  if (!delete_entry_name) {
+    return MZ_FALSE;
+  }
+
+  mz_bool res = MZ_FALSE;
+  if (strcmp(file_name, delete_entry_name) == 0) {
+    res = MZ_TRUE;
+    goto cleanup;
+  }
+
+cleanup:
+  CLEANUP(delete_entry_name);
+  return res;
+}
+
+static int init_entry_mark_array(struct zip_t *zip,
+                                 struct entry_mark *entry_mark_array, int n,
+                                 char *entries[], const int num) {
+  if (!zip || !entry_mark_array || !entries) {
+    return -1;
+  }
+
+  mz_zip_archive_file_stat file_stat;
+  mz_uint64 d_pos = ~0;
+  for (int i = 0; i < n; ++i) {
```

Comment:
"so if we zip_entry_openbyindex, maybe we can (for consistency) just call zip_entry_close after mz_zip_reader_file_stat (instead of CLEANUP here) - does it make sense?" The problem is each time zip_entry_openbyindex() is called, memory allocation for zip->entry.name is generated. If zip_entry_openbyindex() is invoked for multiple times, memory leakage occurs. Before assigning a value to zip->entry.name, check whether the value is NULL. If not, then CLEANUP(zip->entry.name). I'll show it in my next submission.
”I think, if we go from the last to the first one for (int i = n-1; i >= 0; i--) then we can mark DELETE and MOVE and offset in one loop (because we may always keep currently the max offset).
So, in theory we will not need the last loop:” Cause [4.4.1.3 The entries in the central directory MAY NOT necessarily
be in the same order that files appear in the .ZIP file.](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT)，so In the first loop, we need to find the deleted file with the smallest m_local_header_ofs. In the second loop, we need to set the KEEP after the smallest  m_local_header_ofs to MOVE.